File: app/src/main/java/emu/skyline/LogAdapter.java
package emu.skyline;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;

class LogItem extends BaseItem {
    private final String content;
    private final String level;

    LogItem(final String content, final String level) {
        this.content = content;
        this.level = level;
    }

    public String getLevel() {
        return level;
    }

    public String getMessage() {
        return content;
    }

    @Override
    String key() {
        return getMessage();
    }
}

public class LogAdapter extends HeaderAdapter<LogItem> implements View.OnLongClickListener {
    private final ClipboardManager clipboard;
    private final int debug_level;
    private final String[] level_str;
    private final boolean compact;

    LogAdapter(final Context context, final boolean compact, final int debug_level, final String[] level_str) {
        super(context);
        clipboard = (ClipboardManager) context.getSystemService(Context.CLIPBOARD_SERVICE);
        this.debug_level = debug_level;
        this.level_str = level_str;
        this.compact = compact;
    }

    void add(String log_line) {
        try {
            final String[] log_meta = log_line.split("\\|", 3);
            if (log_meta[0].startsWith("1")) {
                final int level = Integer.parseInt(log_meta[1]);
                if (level > debug_level) return;
                add(new LogItem(log_meta[2].replace('\\', '\n'), level_str[level]), ContentType.Item);
            } else {
                add(log_meta[1], ContentType.Header);
            }
        } catch (final IndexOutOfBoundsException ignored) {}
    }

    @Override
    public boolean onLongClick(final View view) {
        final LogItem item = (LogItem) getItem(((LogAdapter.ViewHolder) view.getTag()).position);
        clipboard.setPrimaryClip(ClipData.newPlainText("Log Message", item.getMessage() + " (" + item.getLevel() + ")"));
        Toast.makeText(view.getContext(), "Copied to clipboard", Toast.LENGTH_LONG).show();
        return false;
    }

    @NonNull
    @Override
    public View getView(final int position, View convertView, @NonNull final ViewGroup parent) {
        final LogAdapter.ViewHolder viewHolder;
        final int type = type_array.get(position).type;
        if (convertView == null) {
            viewHolder = new LogAdapter.ViewHolder();
            final LayoutInflater inflater = LayoutInflater.from(HeaderAdapter.mContext);
            if (type == ContentType.Item) {
                if (compact) {
                    convertView = inflater.inflate(R.layout.log_item_compact, parent, false);
                    viewHolder.txtTitle = convertView.findViewById(R.id.text_title);
                } else {
                    convertView = inflater.inflate(R.layout.log_item, parent, false);
                    viewHolder.txtTitle = convertView.findViewById(R.id.text_title);
                    viewHolder.txtSub = convertView.findViewById(R.id.text_subtitle);
                }
                convertView.setOnLongClickListener(this);
            } else {
                convertView = inflater.inflate(R.layout.section_item, parent, false);
                viewHolder.txtTitle = convertView.findViewById(R.id.text_title);
            }
            convertView.setTag(viewHolder);
        } else {
            viewHolder = (LogAdapter.ViewHolder) convertView.getTag();
        }
        if (type == ContentType.Item) {
            final LogItem data = (LogItem) getItem(position);
            viewHolder.txtTitle.setText(data.getMessage());
            if (!compact)
                viewHolder.txtSub.setText(data.getLevel());
        } else {
            viewHolder.txtTitle.setText((String) getItem(position));
        }
        viewHolder.position = position;
        return convertView;
    }

    private static class ViewHolder {
        TextView txtTitle;
        TextView txtSub;
        int position;
    }
}

File: app/src/main/java/emu/skyline/adapter/LogAdapter.kt
/*
 * SPDX-License-Identifier: MPL-2.0
 * Copyright © 2020 Skyline Team and Contributors (https://github.com/skyline-emu/)
 */

package emu.skyline.adapter

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.recyclerview.widget.RecyclerView
import emu.skyline.R
import emu.skyline.data.BaseItem
import kotlinx.android.extensions.LayoutContainer
import kotlinx.android.synthetic.main.log_item.*

/**
 * This class is used to hold all data about a log entry
 */
internal class LogItem(val message : String, val level : String) : BaseItem() {
    /**
     * The log message itself is used as the search key
     */
    override fun key() : String? {
        return message
    }
}

/**
 * This adapter is used for displaying logs outputted by the application
 */
internal class LogAdapter internal constructor(val context : Context, val compact : Boolean, private val debug_level : Int, private val level_str : Array<String>) : HeaderAdapter<LogItem, BaseHeader, RecyclerView.ViewHolder>() {
    private val clipboard : ClipboardManager = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager

    /**
     * This function adds a line to this log adapter
     */
    fun add(logLine : String) {
        try {
            val logMeta = logLine.split("|", limit = 3)

            if (logMeta[0].startsWith("1")) {
                val level = logMeta[1].toInt()
                if (level > debug_level) return

                addItem(LogItem(logMeta[2].replace('\\', '\n'), level_str[level]))
            } else {
                addHeader(BaseHeader(logMeta[1]))
            }
        } catch (ignored : IndexOutOfBoundsException) {
        } catch (ignored : NumberFormatException) {
        }
    }

    private class ItemViewHolder(override val containerView : View) : RecyclerView.ViewHolder(containerView), LayoutContainer

    private class HeaderViewHolder(override val containerView : View) : RecyclerView.ViewHolder(containerView), LayoutContainer

    /**
     * This function creates the view-holder of type [viewType] with the layout parent as [parent]
     */
    override fun onCreateViewHolder(parent : ViewGroup, viewType : Int) : RecyclerView.ViewHolder {
        val inflater = LayoutInflater.from(context)

        val view = when (ElementType.values()[viewType]) {
            ElementType.Item -> inflater.inflate(if (compact) R.layout.log_item_compact else R.layout.log_item, parent, false)

            ElementType.Header -> inflater.inflate(R.layout.log_item, parent, false)
        }

        return when (ElementType.values()[viewType]) {
            ElementType.Item -> {
                if (compact) {
                    ItemViewHolder(view)
                } else {
                    ItemViewHolder(view)
                }
            }

            ElementType.Header -> {
                HeaderViewHolder(view)
            }
        }
    }

    /**
     * This function binds the item at [position] to the supplied [holder]
     */
    override fun onBindViewHolder(holder : RecyclerView.ViewHolder, position : Int) {
        val item = getItem(position)

        if (item is LogItem && holder is ItemViewHolder) {
            holder.text_title.text = item.message
            holder.text_subtitle?.text = item.level

            holder.itemView.setOnClickListener {
                clipboard.setPrimaryClip(ClipData.newPlainText("Log Message", item.message + " (" + item.level + ")"))
                Toast.makeText(holder.itemView.context, "Copied to clipboard", Toast.LENGTH_LONG).show()
            }
        } else if (item is BaseHeader && holder is HeaderViewHolder) {
            holder.text_title.text = item.title
        }
    }
}
